### PR TITLE
G-Modules missing methods (part 1)

### PR DESCRIPTION
**Pull request by mjrodgers**
This is split off from #3266 and should be the "easy" part, this will make way for the rest of the work to be done addressing #3224 .

This should be good to merge

### CODE REVIEW

**Comment by thofma** [comment]
What does `@which x == y` say, when `x` and `y` are GModules?

**Comment by mjrodgers** [comment]
> What does `@which x == y` say, when `x` and `y` are GModules?

I get
```
julia> @which (ME1 == ME2)
==(x, y)
     @ Base Base.jl:165
```
So I guess that means we don't have an implemented method to determine equality?

**Comment by thofma** [comment]
This is the generic `==` fallback, which just calls `===`. You need to implement your own `==` for GModules.

**Comment by fingolfin** [comment]
As a reminder, you implement `==` then you should also implement `hash`

File: experimental/GModule/test/runtests.jl [hunk 1]
```diff
@@ -44,6 +44,29 @@ end
   @test length(Oscar.RepPc.reps(GF(7, 6), G)) == 7
   @test length(Oscar.RepPc.reps(GF(2, 6), G)) == 3
 
+  G = SL(2,3)
+  F = GF(3,2)
+  L = [
+          matrix(F.([0 0 1; 1 0 0; 0 1 0])),
+          matrix(F.([2 0 0; 0 1 0; 0 0 2])),
+          matrix(F.([2 0 0; 0 2 0; 0 0 1])),
+          matrix(F.([1 0 0; 0 1 0; 0 0 1]))
```
Comment (fingolfin) [inline experimental/GModule/test/runtests.jl]: This seems weird, shouldn't it be this (and then similar a few lines later)? I mean both will work, but the current code allocates more.
```suggestion
          matrix(F, [0 0 1; 1 0 0; 0 1 0]),
          matrix(F, [2 0 0; 0 1 0; 0 0 2]),
          matrix(F, [2 0 0; 0 2 0; 0 0 1]),
          matrix(F, [1 0 0; 0 1 0; 0 0 1])
```

File: experimental/GModule/test/runtests.jl [hunk 2]
```diff
@@ -44,6 +44,29 @@ end
...
+  phi = embed(F, E)
+  LE = [
+          matrix(E.([0 0 1; 1 0 0; 0 1 0])),
+          matrix(E.([2 0 0; 0 1 0; 0 0 2])),
+          matrix(E.([2 0 0; 0 2 0; 0 0 1])),
+          matrix(E.([1 0 0; 0 1 0; 0 0 1]))
+      ]
+  mE = free_module(E, 3)
+
+  @test Oscar.GModuleFromGap.extension_of_scalars(M, phi) == GModule(mE, G, [hom(mE, mE, a) for a in LE])
```
Comment (fingolfin) [inline experimental/GModule/test/runtests.jl]: Shouldn't we export `extension_of_scalars`?

File: experimental/GModule/GModule.jl [hunk 3]
```diff
@@ -1563,6 +1572,7 @@ end #module GModuleFromGap
 
 using .GModuleFromGap
 
+export extension_of_scalars
```
Comment (lgoettgens) [inline experimental/GModule/GModule.jl:1681]: this needs to be added to line 1520 as well

**Comment by mjrodgers** [comment]
I've implemented `==` and a hash function for `GModule`, and incorporated some work by @fieker to support functionality needed for the book. There is some work that needs to be done on writing better tests, and in particular implementing a type for GModule homomorphisms, but we feel this can be merged for now and these other things can come as part of another pull request.

**Comment by lgoettgens** [comment]
You need to bump the AA compat entry to 0.40.1 to make sure that the direct_sum stuff from AA is available.

**Comment by fieker** [comment]
... and rebase?

**Comment by lgoettgens** [comment]
`SubfieldLattice_Module.AnticNumberField is deprecated, use Nemo.AbsSimpleNumField instead.`

**Comment by mjrodgers** [comment]
> ... and rebase?

Weird, I did rebase, but it looks like it didn't work. I will redo it (and make the other changes)

**Comment by mjrodgers** [comment]
Okay, rebased (successfully), removed outdated types, and updated the `AbstractAlgebra` dependency

File: experimental/GaloisGrp/src/Subfields.jl [hunk 4]
```diff
@@ -440,4 +479,13 @@ function _subfields(K::AbsSimpleNumField; pStart = 2*degree(K)+1, prime = 0)
   end
 end
 
+function subfield_lattice(K::AbsSimpleNumFiels)
```
Comment (lgoettgens) [inline experimental/GaloisGrp/src/Subfields.jl]:
```suggestion
function subfield_lattice(K::AbsSimpleNumField)
```

**Comment by mjrodgers** [comment]
I'm trying to figure out what's going on with a weird error trying to compute the intersection of two subfield lattice elements, hopefully this will be something minor.

File: experimental/GModule/Brueckner.jl [hunk 5]
```diff
@@ -458,6 +474,11 @@ function lift(C::GModule, mp::Map)
   return allG
 end
 
+function solvable_quotient(G::Oscar.GAPGroup)
+  q = pc_group(symmetric_group(1))
+  mp = hom(G, q, [one(q) for g = gens(G)])
```
Comment (fingolfin) [inline experimental/GModule/Brueckner.jl]: My personal preference
```suggestion
  mp = hom(G, q, [one(q) for g in gens(G)])
```

File: experimental/GModule/Brueckner.jl [hunk 6]
```diff
@@ -176,15 +179,20 @@ the p is in the set.
 function find_primes(mp::Map{<:Oscar.GAPGroup, PcGroup})
   G = domain(mp)
   Q = codomain(mp)
-  I = irreducible_modules(ZZ, Q) 
+  if order(Q) == 1
+    F = free_module(ZZ, 1)
+    I = [gmodule(F, Q, [hom(F, F, [F[1]]) for x = gens(Q)])]
```
Comment (fingolfin) [inline experimental/GModule/Brueckner.jl]:
```suggestion
    I = [gmodule(F, Q, [hom(F, F, [F[1]]) for x in gens(Q)])]
```

File: experimental/GModule/Brueckner.jl [hunk 7]
```diff
@@ -442,11 +451,18 @@ function lift(C::GModule, mp::Map)
...
+    @show [pro[i](epi) for i=1:ngens(G)]
     l= [GMtoGG(reduce(gen(G, i)), pro[i](epi)) for i=1:ngens(G)]
-#    @show map(order, l), order(prod(l))
+    @show map(order, l), order(prod(l))
+    global last_x = l
 #    @show map(order, gens(G)), order(prod(gens(G)))
 
-    h = hom(G, GG, gens(G), l)
+    h = try
+          hom(G, GG, gens(G), l)
```
Comment (fingolfin) [inline experimental/GModule/Brueckner.jl]: I think the generators for the domain can be omitted and then default to `gens(G)`. So does this also work?
```suggestion
          hom(G, GG, l)
```

File: experimental/GModule/Brueckner.jl [hunk 8]
```diff
@@ -458,6 +474,11 @@ function lift(C::GModule, mp::Map)
   return allG
 end
 
+function solvable_quotient(G::Oscar.GAPGroup)
+  q = pc_group(symmetric_group(1))
```
Comment (fingolfin) [inline experimental/GModule/Brueckner.jl]: If you want a trivial pc group, perhaps this is a simpler way
```suggestion
  q = cyclic_group(1)
```

File: experimental/GModule/Cohomology.jl [hunk 9]
```diff
@@ -170,7 +184,8 @@ end
 function fp_group_with_isomorphism(C::GModule)
   #TODO: better for PcGroup!!!
   if !isdefined(C, :F)
-    if order(Group(C)) == 1
+    @show Group(C)
+    if (!isa(Group(C), FPGroup)) && order(Group(C)) == 1
```
Comment (fingolfin) [inline experimental/GModule/Cohomology.jl]: meta-remark: does `group(C)` also work (with lowercase `g`)? Then we should use that.

Also, use `is_trivial(group(C))` ?

File: experimental/GModule/GModule.jl [hunk 10]
```diff
@@ -207,8 +212,71 @@ where `phi` is an embedding of `S` into `R`.
...
+function invariant_lattice_classes(M::GModule{<:Oscar.GAPGroup, <:AbstractAlgebra.FPModule{ZZRingElem}})
+  res = Any[(M, sub(M.M, gens(M.M))[2])]
+  sres = 1
+  new = true
+  lp = keys(factor(order(M.G)).fac)
+  while new
+    new  = false
+    lres = length(res)
+    for X = res[sres:end]
+      for p = lp
```
Comment (fingolfin) [inline experimental/GModule/GModule.jl]: I really don't like the `=` in iteration, would you mind changing it to
```suggestion
    for X in res[sres:end]
      for p in lp
```

File: experimental/GModule/GModule.jl [hunk 11]
```diff
@@ -207,8 +212,71 @@ where `phi` is an embedding of `S` into `R`.
...
+          if any(x->is_isomorphic(r[1], x[1]), res)
+            continue
+          else
+            new = true
+            push!(res, r)
+          end
+        end
+      end
+    end
+    lres = length(res)
```
Comment (fingolfin) [inline experimental/GModule/GModule.jl:256]: Redundant?
```suggestion
```
Also, with this gone, you could also get rid of variable `new` and instead do at the top:
```
  lres = 0
  while lres < length(res)
    lres = length(res)
```

**Review by fingolfin — COMMENTED** [review]
A few minor comments

@mjrodgers what exactly is the error you are getting? Perhaps we can help you debug it.

**Comment by mjrodgers** [comment]
> A few minor comments
> 
> @mjrodgers what exactly is the error you are getting? Perhaps we can help you debug it.

I'm getting the following error from the `intersect` method:
```julia
Zx, x = ZZ["x"]
k, a = number_field(swinnerton_dyer(3, x))
s = subfield_lattice(k)
length(s) == 14

intersect(s[3], s[4])
```
triggers
```
julia> intersect(s[3], s[4])
ERROR: BoundsError: attempt to access 2-element Vector{Vector{Int64}} at index [[3, 4]]
Stacktrace:
 [1] throw_boundserror(A::Vector{Vector{Int64}}, I::Tuple{Vector{Int64}})
   @ Base ./abstractarray.jl:737
 [2] checkbounds
   @ ./abstractarray.jl:702 [inlined]
 [3] _getindex
   @ ./multidimensional.jl:888 [inlined]
 [4] getindex
   @ ./abstractarray.jl:1291 [inlined]
 [5] intersect(A::Oscar.SubfieldLattice_Module.SubfieldLatticeElem{…}, B::Oscar.SubfieldLattice_Module.SubfieldLatticeElem{…})
   @ Oscar.SubfieldLattice_Module ~/.julia/dev/Oscar/experimental/GaloisGrp/src/Subfields.jl:141
 [6] top-level scope
   @ REPL[10]:1
```
Claus made some minor changes to the `intersect` method, but nothing that I can see that would obviously lead to this issue.

File: experimental/GModule/Brueckner.jl (hunk 8 above)
Comment (mjrodgers) [reply experimental/GModule/Brueckner.jl]: These seem to give subtly different groups in some way (`==` returns `false`, even though they are both Pc groups of order 1....), I'll check that this change doesn't cause problems.

File: experimental/GModule/GModule.jl (hunk 11 above)
Comment (mjrodgers) [reply experimental/GModule/GModule.jl:256]: This was Claus' code, I'm supposed to be looking it over to understand and see where improvements can be made; it seems like this change removing `new` leaves everything functioning the same, but I'm double checking everything.

**Comment by mjrodgers** [comment]
Okay, I think I ironed out the bug in `intersect` (it looked like there was a typo and the wrong array was being accessed)

File: experimental/GModule/Brueckner.jl [hunk 12]
```diff
@@ -11,13 +11,16 @@ function Hecke.roots(a::FinFieldElem, i::Int)
   return roots(x^i-a)
 end
 
+Oscar.matrix(phi::Generic.IdentityMap{<:AbstractAlgebra.FPModule}) = identity_matrix(base_ring(domain(phi)), dim(domain(phi)))
```
Comment (fingolfin) [inline experimental/GModule/Brueckner.jl:14]: Should this be added to AbstractAlgebra (not now but down the road)?

File: experimental/GModule/Brueckner.jl [hunk 13]
```diff
@@ -210,9 +218,9 @@ function find_primes(mp::Map{<:Oscar.GAPGroup, PcGroup})
     are correct...)
     TODO: this is not (yet) implemented this way
     =#
-    q = cokernel(b)[1]
+    @show q = cokernel(b)[1]
```
Comment (fingolfin) [inline experimental/GModule/Brueckner.jl]:
```suggestion
    q = cokernel(b)[1]
```

File: experimental/GModule/Brueckner.jl [hunk 14]
```diff
@@ -210,9 +218,9 @@ function find_primes(mp::Map{<:Oscar.GAPGroup, PcGroup})
...
 #    q = quo(kernel(da)[1], image(db)[1])[1]
-    t = torsion_subgroup(q)[1]
+    @show t = torsion_subgroup(q)[1]
```
Comment (fingolfin) [inline experimental/GModule/Brueckner.jl]:
```suggestion
    t = torsion_subgroup(q)[1]
```

File: experimental/GModule/Brueckner.jl [hunk 15]
```diff
@@ -402,6 +410,7 @@ function lift(C::GModule, mp::Map)
   seen = Set{Tuple{elem_type(D), elem_type(codomain(mH2))}}()
   #TODO: the projection maps seem to be rather slow - in particular
   #      as they SHOULD be trivial...
+  global last_k = k
```
Comment (fingolfin) [inline experimental/GModule/Brueckner.jl]:
```suggestion
```

File: experimental/GModule/Brueckner.jl [hunk 16]
```diff
@@ -442,11 +451,18 @@ function lift(C::GModule, mp::Map)
       end
       return d
     end
+    @show [pro[i](epi) for i=1:ngens(G)]
```
Comment (fingolfin) [inline experimental/GModule/Brueckner.jl]:
```suggestion
```

File: experimental/GModule/Brueckner.jl [hunk 17]
```diff
@@ -442,11 +451,18 @@ function lift(C::GModule, mp::Map)
...
     l= [GMtoGG(reduce(gen(G, i)), pro[i](epi)) for i=1:ngens(G)]
-#    @show map(order, l), order(prod(l))
+    @show map(order, l), order(prod(l))
+    global last_x = l
```
Comment (fingolfin) [inline experimental/GModule/Brueckner.jl]:
```suggestion
#    @show map(order, l), order(prod(l))
```

File: experimental/GModule/Brueckner.jl [hunk 18]
```diff
@@ -442,11 +451,18 @@ function lift(C::GModule, mp::Map)
...
 #    @show map(order, gens(G)), order(prod(gens(G)))
 
-    h = hom(G, GG, gens(G), l)
+    h = try
+          hom(G, GG, l)
+        catch
+          @show :crash
+          continue
+        end
```
Comment (fingolfin) [inline experimental/GModule/Brueckner.jl:462]:
```suggestion
    h = hom(G, GG, gens(G), l)
```
or
```suggestion
    h = hom(G, GG, l)
```

File: experimental/GModule/Cohomology.jl [hunk 19]
```diff
@@ -170,19 +184,20 @@ end
 function fp_group_with_isomorphism(C::GModule)
   #TODO: better for PcGroup!!!
   if !isdefined(C, :F)
-    if order(Group(C)) == 1
+    @show Group(C)
```
Comment (fingolfin) [inline experimental/GModule/Cohomology.jl]:
```suggestion
```

File: experimental/GModule/GModule.jl [hunk 20]
```diff
@@ -207,8 +212,71 @@ where `phi` is an embedding of `S` into `R`.
...
+  return res
+end
+
+Oscar.rank(M::AbstractAlgebra.Generic.Submodule{ZZRingElem}) = ngens(M)
+
+function maximal_submodule_bases(M::GModule{<:Oscar.GAPGroup, <:AbstractAlgebra.FPModule{<:FinFieldElem}})
+  C = Gap(M)
+  S = GAP.Globals.MTX.BasesMaximalSubmodules(C)
+  res = []
+  for s = S
```
Comment (fingolfin) [inline experimental/GModule/GModule.jl:267]:
```suggestion
  for s in S
```

File: experimental/GModule/GModule.jl [hunk 21]
```diff
@@ -207,8 +212,71 @@ where `phi` is an embedding of `S` into `R`.
...
+  end
+  return res
+end
+
+Oscar.rank(M::AbstractAlgebra.Generic.Submodule{ZZRingElem}) = ngens(M)
+
+function maximal_submodule_bases(M::GModule{<:Oscar.GAPGroup, <:AbstractAlgebra.FPModule{<:FinFieldElem}})
+  C = Gap(M)
+  S = GAP.Globals.MTX.BasesMaximalSubmodules(C)
+  res = []
```
Comment (fingolfin) [inline experimental/GModule/GModule.jl:266]: This leaves `res` as a `Vector{Any}`, perhaps we can improve that...
```suggestion
  res = dense_matrix_type(base_ring(M))[]
```

File: experimental/GModule/GModule.jl [hunk 22]
```diff
@@ -476,6 +544,10 @@ function Oscar.character_field(C::GModule{<:Any, <:AbstractAlgebra.FPModule{AbsS
   return _character_field(C)[1]
 end
 
+function Oscar.character(C::GModule{<:Any, <:AbstractAlgebra.FPModule{QQAbElem{AbsSimpleNumFieldElem}}})
+  return Oscar.class_function(group(C), [x[2] for x = _character(C)])
```
Comment (fingolfin) [inline experimental/GModule/GModule.jl:548]:
```suggestion
  return Oscar.class_function(group(C), [x[2] for x in _character(C)])
```

File: experimental/GModule/GModule.jl [hunk 23]
```diff
@@ -556,21 +632,21 @@ function gmodule_minimal_field(C::GModule{<:Any, <:AbstractAlgebra.FPModule{<:Fi
 end
 
 function gmodule_minimal_field(C::GModule{<:Any, <:AbstractAlgebra.FPModule{AbsSimpleNumFieldElem}})
-  return _minimize(C) 
+  return _minimize(C)
 end
 
 """
-    gmodule_over(Field)
+    gmodule_over(Field, GModule)
```
Comment (fingolfin) [inline experimental/GModule/GModule.jl:639]:
```suggestion
    gmodule_over(k::Field, C::GModule)
```

File: experimental/GModule/GModule.jl [hunk 24]
```diff
@@ -898,11 +978,17 @@ A needs to have norm 1 wrt. to s, so
 for ord(s) = os = n+1. Then this will find B s.th.
   A = B^(1-s)
 """
-function hilbert90_cyclic(A::MatElem{<:FinFieldElem}, s, os::Int)
+function hilbert90_cyclic(A::MatElem{<:FieldElem}, s, os::Int)
   #apart from rand, this would also work over a number field
+  k= base_ring(A)
```
Comment (fingolfin) [inline experimental/GModule/GModule.jl:983]:
```suggestion
  k = base_ring(A)
```

File: experimental/GModule/GModule.jl [hunk 25]
```diff
@@ -1453,6 +1566,7 @@ end
 
 function Oscar.simplify(C::GModule{<:Any, <:AbstractAlgebra.FPModule{ZZRingElem}})
  f = invariant_forms(C)[1]
+ global last_f = f
```
Comment (fingolfin) [inline experimental/GModule/GModule.jl:1569]:
```suggestion
```

File: experimental/GaloisGrp/src/Subfields.jl [hunk 26]
```diff
@@ -109,6 +110,7 @@ function Base.:*(A::SubfieldLatticeElem, B::SubfieldLatticeElem)
   cs = B.b
   ds = [intersect(b, c) for b = bs for c = cs]
   ds = [x for x = ds if length(x) > 0]
+  sort!(ds)
```
Comment (fingolfin) [inline experimental/GaloisGrp/src/Subfields.jl:113]: Is this necessary, or for debugging, or what? (same for the other `sort!` calls you added)

File: experimental/GModule/GModule.jl [hunk 27]
```diff
@@ -112,7 +112,12 @@ case that `S` is a number field and `R` is the ring of integers in `S`,
 for example `R = ZZ` and `S = QQ`)
```
Comment (fingolfin) [inline experimental/GModule/GModule.jl:112]: Those "examples" seem more like a TODO, perhaps you can also fill them in? (In a follow-up PR, though perhaps at least add "TODO" in this one so it won't be forgotten again.

The docstring above also claims 
> The syntax `M ⊗ phi` is supported.

which doesn't seem to be actually the case, so perhaps another TODO?

**Review by fingolfin — APPROVED** [review]
This seems to be passing tests now. For me it would be OK to merge this now as long as my remaining comments here are not forgotten but addressed in a follow-up PR.

**Comment by fingolfin** [comment]
Thanks for merging, @fieker 

@mjrodgers as mentioned, please take care of the comments here in a follow-up PR :-)

**Comment by benlorenz** [comment]
@fingolfin Do we really need this for the first 1.0 release, is there anything in the book that relies on this?

_Edit:_ was needed for the book and is backported now.

**Comment by mjrodgers** [comment]
I will address them, currently working on the follow-up (and some G-Set documentation)

**Pull request merged**